lib/net.c

```diff
@@ -95,125 +95,187 @@ randomize_rss_key(int guarantee_random_entropy)
-		.src_ip = 0,
-		.src_ip_mask = 0,
-		.dst_port = dst_port_be,
-		.dst_port_mask = dst_port_mask_be,
-		.src_port = src_port_be,
-		.src_port_mask = src_port_mask_be,
-		.proto = proto,
-		.proto_mask = UINT8_MAX,
-		.tcp_flags = 0,
-		.priority = 1,
-		.queue = queue_id,
-	};
+	struct rte_flow_attr attr;
+	struct rte_flow_item pattern[3];
+	struct rte_flow_action action[2];
+	struct rte_flow *flow;
+	struct rte_flow_action_queue queue = { .index = queue_id };
+	struct rte_flow_item_eth eth_spec;
+	struct rte_flow_item_eth eth_mask;
+	struct rte_flow_item_ipv4 ip_spec;
+	struct rte_flow_item_ipv4 ip_mask;
+	struct rte_flow_error error;
+	int ret;
 
-	for (i = 0; i < iface->num_ports; i++) {
-		uint16_t port_id = iface->ports[i];
-		int ret = rte_eth_dev_filter_ctrl(port_id,
-			RTE_ETH_FILTER_NTUPLE,
-			RTE_ETH_FILTER_ADD,
-			&filter_v4);
-		switch (ret) {
-		case 0:
-			/* Success. */
-			break;
-		case -ENOTSUP:
-			G_LOG(ERR,
-				"net: hardware doesn't support adding an IPv4 ntuple filter on port %hhu\n",
-				port_id);
-			return -1;
-		case -ENODEV:
-			G_LOG(ERR,
-				"net: port %hhu is invalid for adding an IPv4 ntuple filter\n",
-				port_id);
-			return -1;
-		default:
-			G_LOG(ERR,
-				"net: implementation-specific error (%d: %s) on port %hhu for adding an IPv4 ntuple filter\n",
-				ret, strerror(-ret), port_id);
-			return -1;
-		}
+	if (!iface->rss) {
+		/*
+		 * IPv4 flows can only be used if supported by the NIC
+		 * (to steer matching packets) and if RSS is supported
+		 * (to steer non-matching packets elsewhere).
+		 */
+		G_LOG(NOTICE, "net: cannot use IPv4 flows when RSS is not supported\n");
+		return -1;
+	}
+
+	memset(&attr, 0, sizeof(struct rte_flow_attr));
+	attr.ingress = 1;
+
+	memset(action, 0, sizeof(action));
+	action[0].type = RTE_FLOW_ACTION_TYPE_QUEUE;
+	action[0].conf = &queue;
+	action[1].type = RTE_FLOW_ACTION_TYPE_END;
+
+	memset(&eth_mask, 0, sizeof(struct rte_flow_item_eth));
+	eth_spec.type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+	eth_mask.type = 0xFFFF;
+	pattern[0].spec = &eth_spec;
+	pattern[0].mask = &eth_mask;
+
+	memset(&ip_spec, 0, sizeof(struct rte_flow_item_ipv4));
+	memset(&ip_mask, 0, sizeof(struct rte_flow_item_ipv4));
+	ip_spec.hdr.dst_addr = dst_ip_be;
+	ip_mask.hdr.dst_addr = 0xFFFFFFFF;
+	ip_spec.hdr.next_proto_id = proto;
+	ip_mask.hdr.next_proto_id = 0xFF;
+	pattern[1].type = RTE_FLOW_ITEM_TYPE_IPV4;
+	pattern[1].spec = &ip_spec;
+	pattern[1].mask = &ip_mask;
+
+	if (proto == IPPROTO_TCP) {
+		struct rte_flow_item_tcp tcp_spec;
+		struct rte_flow_item_tcp tcp_mask;
+		memset(&tcp_spec, 0, sizeof(struct rte_flow_item_tcp));
+		memset(&tcp_mask, 0, sizeof(struct rte_flow_item_tcp));
+		tcp_spec.hdr.src_port = src_port_be;
+		tcp_mask.hdr.src_port = src_port_mask_be;
+		tcp_spec.hdr.dst_port = dst_port_be;
+		tcp_mask.hdr.dst_port = dst_port_mask_be;
+		pattern[2].type = RTE_FLOW_ITEM_TYPE_TCP;
+		pattern[2].spec = &tcp_spec;
+		pattern[2].mask = &tcp_mask;
```

Comment:
Bug: variables `tcp_spec` and `tcp_mask` are only valid inside of this code block, but `pattern` is used later outside of this code block and it carries references to variables `tcp_spec` and `tcp_mask`. Therefore, variables `tcp_spec` and `tcp_mask` must be defined outside of this code block.

This comment applies analogously to variables `udp_spec` and `udp_mask` in the `else if` block below